app/models/user.rb

```diff
@@ -268,10 +268,24 @@ def downcase_email
     self.email.downcase!
   end
 
-  def password_complexity
-    if password.present? and not password.match(/^(?=.*[0-9])(?=.*[A-Za-z])([-a-zA-Z0-9`~\!@#$%\^&*()-_=\+\[\{\]\}\\|;:'",<.>? ]+)$/)
+##
+# Validates the complexity of the password.
+# This method checks if the password meets the specified complexity requirements.
+# If the password is not present or does not meet the requirements, it adds the
+# appropriate error messages to the `errors` object.
+def password_complexity
+  if password.present?
+    uppercase_complexity = Config.password_required_uppercase <= password.gsub(/([^A-Z])/, '').length
+    lowercase_complexity = Config.password_required_lowercase <= password.gsub(/([^a-z])/, '').length
+    numerical_complexity = Config.password_required_numerical <= password.gsub(/([^0-9])/, '').length
+    special_complexity = Config.password_required_special <= password.gsub(/([^\!\@\#\$\%\^\&\*\(\)\-\=\_\+\;\:\<\>\?\~])/, '').length
```

Comment:
I'm sorry, I meant for the code I showed to be a rough example. I don't think these are all the possible special characters. You'll have to add whichever ones I left out